### ISSUE TITLE
bug(router) aux-routes not working with empty-path top-level routes

### ISSUE THREAD

**Issue by choeller** [angular/angular#10726]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Aux-Routes are not found when using two levels of "default routes" (with `path:''`).
This config works:

``` ts
 RouterModule.forRoot([
      {
        path: 'tasks', component: TasksComponent, //container-component has two outlets! 
        children: [
          { path: '', component:TaskListComponent },
          { path: 'overview/:id', component: TasksComponent, outlet: 'right' },
        ]
      }
  ],
```

I can open the Tasks-overview by doing:

`http://localhost:4200/tasks/(right:overview/3)`

See this plnkr and click "to list", "open aux"

As I want to extract the whole Tasks-Functionality into a `TasksModule` I need to be able to do something like this:

``` ts
 RouterModule.forRoot([
      {
        path: 'tasks', : children: [ // this one would be in main module
          //this is defined in the TasksModule:
          { path: '', component: TasksComponent, //container-component has two outlets! 
            children: [
              { path: '', component:TaskListComponent },
              { path: 'overview/:id', component: TasksComponent, outlet: 'right' },
          ]
      }
  ],
```

I would assume both configs are equivalent, and all of the "normal" routing works perfectly fine. But now the link above `http://localhost:4200/tasks/(right:overview/3)` leads to:

`Uncaught (in promise): Error: Cannot match any routes: 'tasks'`

See this plnkr: http://plnkr.co/edit/xd77Mkpqfq6U9tKM4USU?p=preview

**Expected/desired behavior**

Aux-Routes should also work with empty-path parents

**Reproduction of the problem**

Working version: http://plnkr.co/edit/Twx4hXQ0TLfHfrYSe4MO?p=preview
Non-working version:  http://plnkr.co/edit/xd77Mkpqfq6U9tKM4USU?p=preview

**What is the expected behavior?**

Both versions of route definition should lead to the same behavior

**What is the motivation / use case for changing the behavior?**

With the current implementation it is impossible to create an encapsulated Config / Module for the tasks behaviour and include it under a certain path into the main config e.g. "myTasks". 

/cc @vsavkin 

**Please tell us about your environment:**
- **Angular version:** 2.0.0-rc.5
- **Browser:** [all ] 
- **Language:** [TS ]

**Comment by choeller** [comment]
@vsavkin are there plans to fix this? I think the current behaviour makes it impossible to use aux-routes in combination with lazy-loading / feature-module scenarios

**Comment by sambaek** [comment]
Any updates on this? This is blocking us from building out a feature. Would love to see this fixed or a proposed workaround

**Comment by jeanpaulattard** [comment]
I have also encountered this issue, and is blocking me from performing certain actions... What are the plans concerning this issue? Are there any plans of tackling it in the near or distant future?

Thanks in advance!

**Comment by mikechamberlain** [comment]
Yeah this one is really frustrating! I've just spent the last day trying to figure out why my child auxiliary route doesn't work as expected, and have finally tracked it down to this bug. Would be great to have some kind of ETA on this.

In case anyone is interested, here is a Plunkr with the workaround implemented, demonstrating how to load child aux routes both from a routerLink and programatically.

http://plnkr.co/edit/N1BxbCn3ZR8yd0lnHZBU?p=preview

**Comment by uberspeck** [comment]
I'd love to see this fixed soon as well! Burned through a day and a half of beating my head against the wall before giving up and falling back to Bootstrap modals. I do love this router though!

**Comment by dietergeerts** [comment]
This needs to be fixed ASAP!!! Some resolves are common for the main component and the auxiliary one, so they HAVE to be declared one level higher, without a path!

**Comment by asadsahi** [comment]
@IgorMinar @pkozlowski-opensource appologies to attract your attention to this issue. Any ETA on this one?

**Comment by jcimoch** [comment]
I spend hours figuring out why it is not working and then I found this bug. It is still not working on angular 4.0.3 :(

**Comment by marius-oprea** [comment]
4.2.3 with the same bug

**Comment by KshitizGIT** [comment]
Are there any plans to fix this bug ??

**Comment by reppners** [comment]
This renders combination of lazy loading / modular route config and secondary routes unusable :-( of course all empty paths can be filled up with nonsense but thats contrary to readable url's and the concept of empty paths.

**Comment by smoke** [comment]
This is still an issue in 5.0.0, 5.0.1, 5.0.2 :(

I think the following are related or same issues:
 - https://github.com/angular/angular/issues/14662
 - https://github.com/angular/angular/issues/17603
 - https://github.com/angular/angular/issues/18271

With advancement of more and more complicated Angular apps this is becoming more and more common bug people hit, so anyone please fix it.

**Comment by lares83** [comment]
Any update about this issue?

**Comment by maburdi94** [comment]
Yep. It would be great to have this fixed. But it is good to know now that it is a bug, so I can stop losing my mind. 😃

**Comment by jasonaden** [comment]
Increased the priority from `medium` to `critical` so this will get looked at sooner.

**Comment by RoyiNamir** [comment]
The only workaround currently is to not use empty path for lazy loaded modules
 with this, it ill work as expected and you can toggle any outlet you want
See answer with demo in my q and a

https://stackoverflow.com/questions/47857150/angular-how-can-i-activate-aux-route-with-lazy-loaded-module

**Comment by andreyan-andreev** [comment]
* choeller opened this issue on 12 Aug 2016
* ngbot bot added this to the Backlog milestone 19 days ago

**No comment!**

**Comment by trotyl** [comment]
@andreyan-andreev The robot was just introduced recently, there's no ngbot before.

**Comment by nicky-lenaers** [comment]
With the new Ivy renderer being being developed for v6, will this be fixed in v6?

**Comment by luckylooke** [comment]
Angular v6.1.9 and this is still the issue :/
[reproduced here](https://stackblitz.com/edit/angular-routing222-gyofle?embed=1&file=src/app/main/main.routing.module.ts)

**Comment by upondeep** [comment]
I've been tracking this issue everyday for 8 month. The prefix of parent route works but feels so weird.

**Comment by StevenLiekens** [comment]
It doesn't help that the plunkr is broken. :/

**Comment by TGC79** [comment]
It's still a problem in 7.0. 

The fact that using lazy loaded sub modules to fragment your codebase into smaller bundles has been pushed as the correct way to build apps for some time now, more attention should be paid to fixing this issue.

Can we please get some love to @miginmrs PR #26243... and get this fixed finally?

**Comment by TGC79** [comment]
Seems like the changes to fix this in #26243 ended up being reverted in #27516.

@IgorMinar can you reopen this so we can continue to have visibility / history of the issue?

**Comment by TGC79** [comment]
@jasonaden Are you able to reopen this? The issue still persists.

**Comment by DaSchTour** [comment]
Seams that there are 5 duplicates around this. I don't get why this is not fixed since years. Fighting since hours with this problem. 😞

**Comment by monteki** [comment]
The issue still persists[2].

**Comment by lincolnthree** [comment]
Still having this issue in Angular 8.

**Comment by TrevorKarjanis** [comment]
The problem may be that supporting this feature would break the following use case.
```
[{
  path: '',
  children: [{
    path: 'child'
  }]
}, {
  path: '',
  outlet: 'named',
  children: [ ... ]
}]
```
While processing URL segments against the route configurations, the router preemptively adds empty path segments for outlets that are not already part of the URL tree presumably because they should be run. When navigating to `/child`, an empty path segment for the "named" outlet would be added and subsequently processed against the empty path routes at the same level. Normally, the segment would match the empty path with the equivalent outlet. Instead, it would match the empty path for the primary outlet because segments would have to match empty paths of any outlet in order to support the use case identified in this issue. The `child` segment would also match the empty path for the primary outlet, so the outlet uniqueness check would fail with the following error (packages/router/src/recognize.ts line 213).

`Two segments cannot have the same outlet name: '' and ''.`

In other words, two routes cannot occupy the same outlet.

https://github.com/angular/angular/compare/master...TrevorKarjanis:secondary-route-support

**Comment by magnolemos** [comment]
is possible by passing the ralativeTo: this.activatedRoute or this.activatedRoute.parent

For example: 

   this.router.navigate([{ outlets: { nameOutlet: ['urlOutlet'] } }], {
      relativeTo: this.activatedRoute.parent
    });

**Comment by atscott** [comment]
Reopening because the fix was reverted:
> Seems like the changes to fix this in #26243 ended up being reverted in #27516.

**Comment by atscott** [comment]
Reproduction in latest Angular: https://stackblitz.com/edit/angular-ivy-7nkhpm?file=src%2Fapp%2Fmain.module.ts

**Comment by PaulBaugher** [comment]
Still?

**Comment by uloga** [comment]
Yep, still! What a disaster the whole project. Known issue with every upgrade/update the same issue appears. :(

**Comment by DoloMike** [comment]
This fixed the issue I was having:

https://stackoverflow.com/a/48831904/2909562

**Comment by jcimoch** [comment]
Sooner than covid-19 vaccine, I can't believie it!

**Comment by Bulforce** [comment]
:star::star::star::star::star: 
:clap::clap::clap::clap::clap:

**Comment by henriacle** [comment]
Finally, gonna try it tomorrow 2021 is already a promising year 🥳🤣

**Comment by atscott** [comment]
This is was included in the `11.0.6` release today so if you're able to update your applications, please do and give it a go.

**Comment by acohenOT** [comment]
Works great. Thank you @atscott !

**Comment by LukasMachetanz** [comment]
I am not sure, but was with this MR maybe another bug introduced. It seems that `router.isActive` does not provide the correct result for auxiliary routes with empty-path top level routes.

I tried it here: https://stackblitz.com/edit/lukmac-angular-router

I think the example should be self explanatory. I would expect the console output to show true and not false.

Can maybe someone confirm the issue and tell if it is somehow corresponding to this. Thanks.
